### PR TITLE
Optimizations branch using for loops rather than map/reduce/function based looping

### PR DESCRIPTION
**Pull request by dbkaplun**
#3 is included in this branch. I didn't include this in the other PR because the performance improvements weren't that great in comparison.

### CODE REVIEW

**Comment by sindresorhus** [comment]
How's the performance improvement with this PR?

File: index.js [hunk 1]
```diff
@@ -13,56 +13,51 @@ module.exports = function (str, opts) {
...
 	var splits = [];
-
-	for (var i = 0; i < stats.length; i++) {
-		// found the separator on the same index on all lines
-		if (stats[i] === lines.length) {
-			splits.push(i);
+	var consecutive = false;
+	for (var col = 0, count; col < counts.length; col++) {
+		var count = counts[col];
+		if (count !== lines.length) consecutive = false;
```
Comment (sindresorhus) [inline index.js]: braces

File: index.js [hunk 2]
```diff
@@ -13,56 +13,51 @@ module.exports = function (str, opts) {
...
-		});
-	});
-
-	var headers = lines.shift();
-
-	if (opts.headers) {
-		headers = opts.headers;
+	var rows = [];
+	var headers = opts.headers;
+	for (var i = 0, line; i < lines.length; i++) {
```
Comment (sindresorhus) [inline index.js]: You have a redefinition of `i` here. It's already used earlier.

Comment (sindresorhus) [inline index.js]: I think the main function could be split into some smaller ones.

**Comment by sindresorhus** [comment]
Awesome!

File: index.js (hunk 1 above)
Comment (dbkaplun) [inline index.js]: Fixed [here](https://github.com/sindresorhus/parse-columns/commit/f415229ecc07d73366802fdd548f2dadacc10b1e#diff-168726dbe96b3ce427e7fedce31bb0bcR44).

File: index.js (hunk 2 above)
Comment (dbkaplun) [inline index.js]: Fixed [here](https://github.com/sindresorhus/parse-columns/commit/f415229ecc07d73366802fdd548f2dadacc10b1e#diff-168726dbe96b3ce427e7fedce31bb0bcR59).

**Comment by dbkaplun** [comment]
Added a benchmark:

```
 10:40:58  dan@elise:~/stuff/parse-columns   asdf ✘ ✚ 
$ git reset --hard optimizations-for
HEAD is now at 2ee4bb2 Add benchmark test case

 10:41:05  dan@elise:~/stuff/parse-columns   asdf ✔ 
$ npm test

> parse-columns@1.0.1 test /Users/dan/stuff/parse-columns
> node test.js

  ✔ parse
  ✔ headers option
  ✔ transform option
  ✔ separator option
  ✔ differing line lengths
10 iterations: 0.2287s
  ✔ benchmark

  ✔ 6 tests passed


 10:41:49  dan@elise:~/stuff/parse-columns   asdf ✔ 
$ git reset --hard HEAD^ && git checkout optimizations-for -- fixtures/lsof.out test.js && npm test
HEAD is now at f415229 Create Table prototype

> parse-columns@1.0.1 test /Users/dan/stuff/parse-columns
> node test.js

  ✔ parse
  ✔ headers option
  ✔ transform option
  ✔ separator option
  ✔ differing line lengths
10 iterations: 0.2029s
  ✔ benchmark

  ✔ 6 tests passed


 10:42:07  dan@elise:~/stuff/parse-columns   asdf ✘ ✚ 
$ git reset --hard HEAD^ && git checkout optimizations-for -- fixtures/lsof.out test.js && npm test
HEAD is now at c4f689b Use for loops instead of map/reduce

> parse-columns@1.0.1 test /Users/dan/stuff/parse-columns
> node test.js

  ✔ parse
  ✔ headers option
  ✔ transform option
  ✔ separator option
  ✔ differing line lengths
10 iterations: 0.1895s
  ✔ benchmark

  ✔ 6 tests passed


 10:42:15  dan@elise:~/stuff/parse-columns   asdf ✘ ✚ 
$ git reset --hard HEAD^ && git checkout optimizations-for -- fixtures/lsof.out test.js && npm test
HEAD is now at 1b4b272 Consolidate another two loops into one

> parse-columns@1.0.1 test /Users/dan/stuff/parse-columns
> node test.js

  ✔ parse
  ✔ headers option
  ✔ transform option
  ✔ separator option
  ✔ differing line lengths
10 iterations: 0.2661s
  ✔ benchmark

  ✔ 6 tests passed


 10:42:23  dan@elise:~/stuff/parse-columns   asdf ✘ ✚ 
$ git reset --hard HEAD^ && git checkout optimizations-for -- fixtures/lsof.out test.js && npm test
HEAD is now at e63db4f Consolidate two loops into one

> parse-columns@1.0.1 test /Users/dan/stuff/parse-columns
> node test.js

  ✔ parse
  ✔ headers option
  ✔ transform option
  ✔ separator option
  ✔ differing line lengths
10 iterations: 0.2112s
  ✔ benchmark

  ✔ 6 tests passed


 10:42:34  dan@elise:~/stuff/parse-columns   asdf ✘ ✚ 
$ git reset --hard HEAD^ && git checkout optimizations-for -- fixtures/lsof.out test.js && npm test
HEAD is now at f9fa6e4 1.0.1

> parse-columns@1.0.1 test /Users/dan/stuff/parse-columns
> node test.js

  ✔ parse
  ✔ headers option
  ✔ transform option
  ✔ separator option
  ✔ differing line lengths
10 iterations: 0.2504s
  ✔ benchmark

  ✔ 6 tests passed
```

**Comment by sindresorhus** [comment]
Don't really see the point in making it a prototype. Meant that it could just be some normal functions.

**Comment by dbkaplun** [comment]
Fixed [here](https://github.com/sindresorhus/parse-columns/commit/e168821711f42a9940e95efcca2ff0e4ec5c5435).

**Comment by dbkaplun** [comment]
Fixed [here](https://github.com/dbkaplun/parse-columns/commit/0f27c71c78f9cf786b83014026cdb084139593b6).

**Comment by sindresorhus** [comment]
Awesome! Thanks so much for this :)

**Pull request closed without merging**